src/ipc/ipc4/dai.c

```diff
@@ -240,6 +240,58 @@ void dai_dma_release(struct comp_dev *dev)
 	}
 }
 
+static int dai_init_llp_info(struct comp_dev *dev)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct ipc4_copier_module_cfg *copier_cfg;
+	union ipc4_connector_node_id node;
+	uint32_t llp_slot_size;
+	uint32_t max_slot;
+	uint32_t node_id;
+	uint32_t offset;
+	int i;
+
+	copier_cfg = dd->dai_spec_config;
+	node = copier_cfg->gtw_cfg.node_id;
+
+	/* HDA doesn't use llp slot */
+	if (dd->ipc_config.type == SOF_DAI_INTEL_HDA)
+		return 0;
+
+	/* don't support more gateway like EVAD */
+	if (node.f.dma_type >= ipc4_max_connector_node_id_type) {
+		comp_err(dev, "unsupported gateway %d", (int)node.f.dma_type);
+		return -EINVAL;
+	}
+
+	dd->slot_info.node_id = node.dw & IPC4_NODE_ID_MASK;
+	llp_slot_size = sizeof(struct ipc4_llp_reading_slot);
+
+	/* sdw with multiple gateways uses sndw_reading_slots */
+	if (node.f.dma_type == ipc4_alh_link_output_class && node.f.v_index >= ALH_MULTI_GTW_BASE) {
+		offset = SRAM_REG_LLP_SNDW_READING_SLOTS;
+		max_slot = IPC4_MAX_LLP_SNDW_READING_SLOTS - 1;
+	} else {
+		offset = SRAM_REG_LLP_GPDMA_READING_SLOTS;
+		max_slot = IPC4_MAX_LLP_GPDMA_READING_SLOTS;
+	}
+
+	for (i = 0; i < max_slot; i++, offset += llp_slot_size) {
+		node_id = mailbox_sw_reg_read(offset + i * llp_slot_size);
+		if (!node_id)
+			break;
+	}
+
+	if (i >= max_slot) {
+		comp_err(dev, "can't find free slot");
+		return -EIO;
```

Comment:
-EINVAL do we also need to set an IPC4 error value ?

Reply:
changed.  We don't need to use ipc4 error at here, it will be converted to ipc4 error in ipc4/handler.c or ipc4/help.c